private-templates-service/client/src/components/Designer/SaveModal/SaveModal.tsx

```diff
@@ -0,0 +1,149 @@
+import * as AdaptiveCards from "adaptivecards";
+import Tags from '../../Common/Tags';
+
+import { Container, TemplateName, ACWrapper, TemplateFooterWrapper, TemplateStateWrapper } from '../../AdaptiveCardPanel/styled';
+import { 
+  BackDrop,
+  Modal, 
+  TitleWrapper, 
+  ColumnWrapper, 
+  InfoWrapper,
+  ButtonWrapper, 
+  MiddleRowWrapper, 
+  Card, 
+  StyledTitle, 
+  StyledH3, 
+  TagsWrapper, 
+  StyledCancelButton, 
+  StyledSaveButton, 
+  StyledTextField } from './styled';
+
+const mapStateToProps = (state: RootState) => {
+  return {
+    templateID: state.currentTemplate.templateID,
+    templateJSON: state.currentTemplate.templateJSON,
+    templateName: state.currentTemplate.templateName,
+    sampleDataJSON: state.currentTemplate.sampleDataJSON,
+    version: state.currentTemplate.version
+  }
+}
+
+interface Props {
+  templateID?: string;
+  templateName?: string;
+  sampleDataJSON?: object;
+  templateJSON?: object;
+  designerSampleData?: any;
+  designerTemplateJSON?: any;
+  version?: string;
+  closeModal: () => void;
+  updateTemplate: (templateID?: string, currentVersion?: string, templateJSON?: object, sampleDataJSON?: object, templateName?: string, state?: PostedTemplate.StateEnum, tags?: string[], isShareable?: boolean ) => any; 
+}
+
+interface State {
+  tags: string[];
+  templateName: string;
+}
+
+const mapDispatchToProps = (dispatch: any) => {
+  return { 
+    closeModal: () => {
+      dispatch(closeModal());
+    },
+    updateTemplate: (templateID?: string, currentVersion?: string, templateJSON?: object, sampleDataJSON?: object, templateName?: string, templateState?: PostedTemplate.StateEnum, templateTags?:string[]) => {
+      dispatch(updateTemplate(templateID, currentVersion, templateJSON, sampleDataJSON, templateName, templateState, templateTags));
+    } 
+  }
+}
+
+class SaveModal extends React.Component<Props,State> {
+  constructor(props: Props) {
+    super(props);
+    this.state = {tags : [], templateName: "Untitled Template"}
+  }
+
+  saveTags = (tagsToUpdate: string[]) => {
+    this.setState({tags: tagsToUpdate});
+  }
+
+  tagRemove = (tag: string) => {
+    const newTags = this.state.tags.filter((existingTag: string) => existingTag !== tag);
+    this.setState({tags: newTags});
+  }
+
+  onChange = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
+    if(newValue){
+      this.setState({templateName: newValue});
+    }
+  }
+
+  onClick = () => {
+    // will only trigger on first save
+    if(JSON.stringify(this.props.sampleDataJSON) !== JSON.stringify(this.props.designerTemplateJSON) || this.props.sampleDataJSON !== this.props.designerSampleData ){
+      this.props.updateTemplate(undefined,undefined,this.props.designerTemplateJSON,this.props.designerSampleData,this.state.templateName,PostedTemplate.StateEnum.Draft,this.state.tags);
+    }
+    this.props.closeModal();
+  }
+  
+  render(){ 
+    var adaptiveCard = new AdaptiveCards.AdaptiveCard();
```

Comment:
Can line 97-102 and the renderingSetup method in AdaptiveCard.tsx be refactored into a separate method in TemplateUtil because they seem to be extremely similar? Would reduce repetitive code.